src/main/java/io/vavr/collection/Iterator.java

```diff
@@ -1722,10 +1722,16 @@ public U next() {
         if (!hasNext()) {
             return Tuple.of(empty(), empty());
         } else {
-            final Stream<T> that = Stream.ofAll(this);
-            final Iterator<T> first = that.iterator().filter(predicate);
-            final Iterator<T> second = that.iterator().filter(predicate.negate());
-            return Tuple.of(first, second);
+            final java.util.List<T> first = new ArrayList<>();
```

Comment:
I do not know if it is an issue here, but...
This solution is filtering eagerly and previous solution was filtering lazily.

If we would like to keep the same behaviour there is an option of creating a memoized predicate and then use it during filtering, e.g.
```java
final Predicate<? super T> memoizedPredicate = Function1.of(predicate::test).memoized()::apply;
final Iterator<T> first = that.iterator().filter(memoizedPredicate);
final Iterator<T> second = that.iterator().filter(memoizedPredicate.negate());
```

Reply:
I found that my solution with memoized predicate will not work in case when a mutable object mutates to the value that corresponds to another object to be filtered. Then another object is not evaluated but taken from cache with wrong result for it.
Example test:
```java
final HashSet<MutableInteger> integers = HashSet.of(MutableInteger.of(1), MutableInteger.of(2));
final Tuple2<HashSet<MutableInteger>, HashSet<MutableInteger>> partition = integers.partition(mutableInteger -> mutableInteger.incrementAndCheckGreaterThan(2));
assertThat(partition._1).isEqualTo(HashSet.of(MutableInteger.of(3)));
assertThat(partition._2).isEqualTo(HashSet.of(MutableInteger.of(2)));
```
Where 
```java
    private static class MutableInteger {
        private int integer;
...
        boolean incrementAndCheckGreaterThan(int i) {
            return integer++ > i;
        }
...
    }
```
```
In this case the result of evaluation `MutableInteger.of(1).incrementAndCheckGreaterThan(2)` is put into cache with entry `MutableInteger.of(2), false`.
Then when the next object from set `MutableInteger.of(2)` is filtered and it is found in the cache and evaluated to `false`. While normally it would evaluate to `MutableInteger.of(3) -> true.


But the question stays whether we want to keep the partitioning lazy or not.

Reply:
This is a bit too technical for me... Maybe @danieldietrich can have an answer about it.